src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs

```diff
@@ -1796,27 +1796,30 @@ protected override void AfterMembersCompletedChecks(BindingDiagnosticBag diagnos
             Debug.Assert(ObsoleteKind != ObsoleteAttributeKind.Uninitialized);
             Debug.Assert(GetMembers().All(m => m.ObsoleteKind != ObsoleteAttributeKind.Uninitialized));
 
-            if (ObsoleteKind != ObsoleteAttributeKind.None
-                || GetMembers().All(m => m is not MethodSymbol { MethodKind: MethodKind.Constructor, ObsoleteKind: ObsoleteAttributeKind.None } method
+            if (ObsoleteKind == ObsoleteAttributeKind.None
+                && !GetMembers().All(m => m is not MethodSymbol { MethodKind: MethodKind.Constructor, ObsoleteKind: ObsoleteAttributeKind.None } method
                                          || !method.ShouldCheckRequiredMembers()))
             {
-                return;
-            }
-
-            foreach (var member in GetMembers())
-            {
-                if (!member.IsRequired())
+                foreach (var member in GetMembers())
                 {
-                    continue;
-                }
+                    if (!member.IsRequired())
+                    {
+                        continue;
+                    }
 
-                if (member.ObsoleteKind != ObsoleteAttributeKind.None)
-                {
-                    // Required member '{0}' should not be attributed with 'ObsoleteAttribute' unless the containing type is obsolete or all constructors are obsolete.
-                    diagnostics.Add(ErrorCode.WRN_ObsoleteMembersShouldNotBeRequired, member.GetFirstLocation(), member);
+                    if (member.ObsoleteKind != ObsoleteAttributeKind.None)
+                    {
+                        // Required member '{0}' should not be attributed with 'ObsoleteAttribute' unless the containing type is obsolete or all constructors are obsolete.
+                        diagnostics.Add(ErrorCode.WRN_ObsoleteMembersShouldNotBeRequired, member.GetFirstLocation(), member);
+                    }
                 }
             }
 
+            if (Indexers.FirstOrDefault() is PropertySymbol indexerSymbol)
+            {
+                Binder.GetWellKnownTypeMember(DeclaringCompilation, WellKnownMember.System_Reflection_DefaultMemberAttribute__ctor, diagnostics, indexerSymbol.TryGetFirstLocation() ?? GetFirstLocation());
```

Comment:
>indexerSymbol.TryGetFirstLocation() ?? GetFirstLocation() [](http://example.com/codeflow?start=145&length=57)

nit: I didn't follow the intricacy of the `Location` logic. Do we have a test where `indexerSymbol.TryGetFirstLocation()` is null?

Reply:
This is defense in depth, theoretically the API can return null, even though we don't expect it to. This pattern is also used elsewhere in this function.